File: find-first-palindromic-string-in-the-array/solution2108.go
// https://leetcode.com/problems/find-first-palindromic-string-in-the-array/

package main

import "fmt"

// T: O(nm), S: O(1)
func firstPalindrome(words []string) string {
	for _, w := range words {
		if isPalindrome(w) {
			return w
		}
	}
	return ""
}

func isPalindrome(word string) bool {
	for i, j := 0, len(word)-1; i < j; i, j = i+1, j-1 {
		if word[i] != word[j] {
			return false
		}
	}
	return true
}

func main() {
	var words1 = []string{"abc", "car", "ada", "racecar", "cool"}
	var words2 = []string{"notapalindrome", "racecar"}
	var words3 = []string{"def", "ghi"}
	fmt.Println(firstPalindrome(words1)) // "ada"
	fmt.Println(firstPalindrome(words2)) // "racecar"
	fmt.Println(firstPalindrome(words3)) // ""
}

File: find-first-palindromic-string-in-the-array/Solution2108.java
// https://leetcode.com/problems/find-first-palindromic-string-in-the-array/

class Solution2108 {

    // T: O(nm), S: O(1)
    public String firstPalindrome(String[] words) {
        for (var w : words) {
            if (isPalindrome(w)) {
                return w;
            }
        }
        return "";
    }

    private boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    public static void main(String[] args) {
        var sol = new Solution2108();
        var words1 = new String[] { "abc", "car", "ada", "racecar", "cool" };
        var words2 = new String[] { "notapalindrome", "racecar" };
        var words3 = new String[] { "def", "ghi" };
        System.out.println(sol.firstPalindrome(words1)); // "ada"
        System.out.println(sol.firstPalindrome(words2)); // "racecar"
        System.out.println(sol.firstPalindrome(words3)); // ""
    }

}
